zeppelin-zengine/src/main/java/org/apache/zeppelin/search/LuceneSearch.java

```diff
@@ -371,6 +372,7 @@ private void deleteDoc(String noteId, Paragraph p) {
    * @see org.apache.zeppelin.search.Search#close()
    */
   @Override
+  @PreDestroy
   public void close() {
     try {
       indexWriter.close();
```

Comment:
This is an overridden method, is it okay if the parent class's `close()` is not called?

Reply:
You are right, we should call `super.close()` to close the LuceneSearch thread.
What is the right place for the `@PreDestroy` annotation?

Reply:
First, it is necessary to check whether `LuceneSearch` instance is created using the Java EE container(dependency injection) in the test you mentioned.

For example, in `NotebookSearviceTest`, `LuceneSearch` instance is created manually, so if the error you mentioned is an error in `NotebookSearviceTest`, adding the `@PreDestroy` annotation will have no effect.

Reply:
> First, it is necessary to check whether `LuceneSearch` instance is created using the Java EE container(dependency injection) in the test you mentioned.

LuceneSearch is started by Jetty - [Link](https://github.com/apache/zeppelin/blob/91bfbf9db1b2ca915644c264694e0add03a69e61/zeppelin-server/src/main/java/org/apache/zeppelin/server/ZeppelinServer.java#L219)
During Zeppelin Server tests, e.g. in [NotebookRestApiTest](https://github.com/apache/zeppelin/blob/master/zeppelin-server/src/test/java/org/apache/zeppelin/rest/NotebookRestApiTest.java), the server is restarted several times.
You can search for the Stacktraces (keyword `LockObtainFailedException`) in the [last CI](https://pipelines.actions.githubusercontent.com/Y581u358DfamG5WExglSJr5GUw71gpY3tFyZWVl6aGaGrt8Y5v/_apis/pipelines/1/runs/1914/signedlogcontent/12?urlExpires=2021-05-06T06%3A22%3A05.5237793Z&urlSigningMethod=HMACV1&urlSignature=hA3ohf2B2%2FYMqxBz3E4eEf8Xgf6QTy75HAaZrmCqouU%3D) run.

You are right, we should also close LuceneSearch in `NotebookSearviceTest`.

Reply:
It seems more appropriate to put `@PreDestroy` in `SearchService::close()`.
And `SearchService::close()` is being called from `NoteBook::close()`, wouldn't it be better to delete it?

Reply:
I think it would be better if Notebook and SearchService worked independently.

I have moved `@PreDestroy` to `SearchService` because this service is started by Jetty.

Reply:
It seems that the annotation `@PreDestroy` is not inherited by `LuceneSearch`. Do you know if inheritance works with `@PreDestroy`?

Reply:
The annotation [`@PreDestroy`](https://docs.oracle.com/javaee/7/api/javax/annotation/PreDestroy.html) doesn't have the annotation [`@Inherited`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/annotation/Inherited.html), so the inheritance does not work. I will undo this change.

Reply:
> 
> 
> The annotation [`@PreDestroy`](https://docs.oracle.com/javaee/7/api/javax/annotation/PreDestroy.html) doesn't have the annotation [`@Inherited`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/annotation/Inherited.html), so the inheritance does not work. I will undo this change.

Perhaps the behavior is different depending on the implementation. When tested in `spring` or `quarkus`, it works by inheriting `@PreDestroy` of the parent class.

Reply:
The `@PreDestroy` annotation works in the child class if you do not overwrite the method with the `@PreDestroy` annotation in the child class.